deploy nodes with terraform

alexandre-allard opened the pull request:
**Component**: terraform

**Context**: Metalk8s deployment with terraform

**Summary**: Add possiblity to deploy nodes using terraform (for now only bootstrap, but will add the expansion really soon)

**Acceptance criteria**: Nothing is broken ?
Ability to deploy a cluster with the bootstrap node installed:
`terraform apply -auto-approve -var=metalk8s_iso=/path/to/metalk8s.iso`

alexandre-allard committed:
tf: create control plane network & interfaces

Creation of the network used by the control plane,
this allows the usage of a VIP for apiserver which
will be used to test backup/restoration mechanism

alexandre-allard committed:
tf: add script to configure control plane iface

This script creates the configuration of the new
network interface used by the control plane network
and ensures that this interface is up.
It would be better to replace it by a cloud-init
script at some point, but for now, it does the job.

alexandre-allard committed:
tf: add the control plane port to the nodes

Link the control plane interfaces to the nodes

alexandre-allard committed:
tf: retrieve networks dynamically in bootstrap config

Networks for workload and control plane were hardcoded
in the bootstrap configuration generation script,
we now retrieve them from configured network interfaces
on the host.
Also enable keepalived, since we now have a VIP.

alexandre-allard committed:
tf: add resources to install the bootstrap node

Add the possiblity to deploy metalk8s on the
bootstrap node depending on an optional variable
`metalk8s_iso`, if defined, it'll try to do it.
Usage:
  terraform apply -auto-approve \
    -var='metalk8s_iso=/path/to/your/iso'

It's also possible to specify the destination of
the ISO `metalk8s_iso_dest` and the mountpoint
`metalk8s_iso_mountpoint` on the remote host.

To only deploy a bootstrap node, it's possible
to do:
  terraform apply -auto-approve \
    -var='metalk8s_iso=/path/to/your/iso' \
    -target=null_resource.metalk8s_run_bootstrap

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/improvement/deploy-metalk8s-with-terraform` with contents from `w/2.3/improvement/deploy-metalk8s-with-terraform`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/improvement/deploy-metalk8s-with-terraform origin/development/2.4
 $ git merge origin/w/2.3/improvement/deploy-metalk8s-with-terraform
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/deploy-metalk8s-with-terraform
```

NicolasT commented on eve/workers/openstack-multiple-nodes/terraform/scripts/network-iface-config.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Can't we do all of this through `cloud-init`?

alexandre-allard replied:
Yes, that's exactly what the comment below says, I don't know yet how to do it through terraform, but I'll search.
The script is more to make the rest of the stuff work until I found the right solution.

NicolasT commented on eve/workers/openstack-multiple-nodes/terraform/network.tf:
```diff
@@ -94,3 +104,87 @@ resource "openstack_networking_secgroup_v2" "bastion" {
+    subnet_id  = openstack_networking_subnet_v2.control_plane.id
+  }
+
+  allowed_address_pairs {
```
Instead of all this mockery, if we have a private control-plane network, can't we simply get rid of port-security on it, create a port for the VIP so it's not being handed out by DHCP, and get rid of IPIP?

alexandre-allard-scality replied:
We can already get rid of IPIP with the current changes, but I'll try so simplify this as you suggested.

alexandre-allard-scality replied:
I tried to disable port security, but I never managed to make it work.
Every time, I end up with some (not really useful) errors when spawning instances, like this one:
```
Error: Error waiting for instance (37abba31-95c7-47df-8be5-bbf1ae8e6033) to become ready: unexpected state 'ERROR', wanted target 'ACTIVE'. last error: %!s(<nil>)
```
I'm still investigating... but I don't know if it worth spending too much time.
Still, IPIP can be removed with these changes.

NicolasT commented on eve/workers/openstack-multiple-nodes/terraform/network.tf:
```diff
@@ -6,6 +6,17 @@ variable "openstack_network" {
+  control_plane_network = {
+    name    = "${local.prefix}-control-plane"
+    cidr    = "192.168.1.0/28"
+    gateway = "192.168.1.14"
```
It's counter-intuitive to have the gateway address as the last address (before broadcast) in a network, in general it'd be the first, but sure this works :man_shrugging:

alexandre-allard replied:
It's only a detail, but I'll change it if it bother you.

alexandre-allard committed:
tf: add bootstrap volume provisioning

Call volume provisioning script in terraform,
if the bootstrap has been run through it.
Step can be disabled adding the following option
on the CLI:
  terraform apply -var=metalk8s_provision_volumes=false

bert-e commented:
# History mismatch



Merge commit #594e2f9f9477bcb1a643d30ac61775817a373391 on the integration branch
`w/2.1/improvement/deploy-metalk8s-with-terraform` is merging a branch which is neither the current
branch `improvement/deploy-metalk8s-with-terraform` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/deploy-metalk8s-with-terraform` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
tf: handle bastion ssh key in terraform

Use terraform provisioner to generate
and copy the bastion SSH keys on every
nodes instead of doing it in the CI.

gdemonet commented on eve/workers/openstack-multiple-nodes/terraform/network.tf:
```diff
@@ -94,3 +104,87 @@ resource "openstack_networking_secgroup_v2" "bastion" {
+  name        = "${local.prefix}-control-plane"
+  network_id  = openstack_networking_network_v2.control_plane.id
+  enable_dhcp = true
+  cidr        = "${local.control_plane_network.cidr}"
```
Minor: now that we use Terraform 0.12, you can just write:

```
  cidr        = local.control_plane_network.cidr
```

Same for other occurences below

alexandre-allard replied:
Nice, will do.

gdemonet commented on eve/workers/openstack-multiple-nodes/terraform/metalk8s.tf:
```diff
@@ -0,0 +1,60 @@
+variable "metalk8s_iso" {
+  type        = string
+  description = "Metalk8s ISO file path to upload to bootstrap node"
+  default     = ""
+}
+
+variable "metalk8s_iso_dest" {
+  type        = string
+  description = "Location where the Metalk8s ISO will be copied on the bootstrap node"
+  default     = "/home/centos/metalk8s.iso"
+}
+
+variable "metalk8s_iso_mountpoint" {
+  type        = string
+  description = "Mountpoint for the Metalk8s ISO"
+  default     = "/var/tmp/metalk8s"
+}
```
Neat! Even if we may not really need it in the CI, it will be helpful for sharing with users :+1:

alexandre-allard replied:
Yep, that's the goal

gdemonet commented on eve/workers/openstack-multiple-nodes/terraform/metalk8s.tf:
```diff
@@ -58,3 +58,34 @@ resource "null_resource" "metalk8s_run_bootstrap" {
   }
 }
+
+variable "metalk8s_provision_volumes" {
+  type    = bool
+  default = false
+}
```
For future improvement: we could (should?) mount an OpenStack volume to store `SparseLoopDevice`s backing files, and maybe let users choose the total size of this volume.

alexandre-allard replied:
OpenStack volume ? You want to bring flakies ?
Yes, could be a good idea.

gdemonet commented on eve/workers/openstack-multiple-nodes/terraform/ssh.tf:
```diff
@@ -46,3 +50,10 @@ resource "null_resource" "ssh_config" {
   }
 }
+
+# Generate Bastion SSH keypair
+resource "null_resource" "bastion_ssh_keys" {
+  provisioner "local-exec" {
+    command  = "ssh-keygen -t rsa -b 4096 -N '' -f ${local.bastion_ssh_key}"
+  }
+}
```
Didn't want to generate it on the host, but yeah, makes sense to do it that way.

gdemonet reviewed (dismissed):
Just a few minor nitpicks, most of it looks good to me! I'll make sure to use these recent changes in my PR as well, let's merge this one :)

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/improvement/deploy-metalk8s-with-terraform` with contents from `w/2.3/improvement/deploy-metalk8s-with-terraform`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/improvement/deploy-metalk8s-with-terraform origin/development/2.4
 $ git merge origin/w/2.3/improvement/deploy-metalk8s-with-terraform
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/deploy-metalk8s-with-terraform
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/deploy-metalk8s-with-terraform` into
`development/2.0`
* [w/2.1/improvement/deploy-metalk8s-with-terraform](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/deploy-metalk8s-with-terraform) will be merged into `development/2.1` 
* [w/2.2/improvement/deploy-metalk8s-with-terraform](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/deploy-metalk8s-with-terraform) will be merged into `development/2.2` 
* [w/2.3/improvement/deploy-metalk8s-with-terraform](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/deploy-metalk8s-with-terraform) will be merged into `development/2.3` 
* [w/2.4/improvement/deploy-metalk8s-with-terraform](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/deploy-metalk8s-with-terraform) will be merged into `development/2.4` 
* [w/2.5/improvement/deploy-metalk8s-with-terraform](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/deploy-metalk8s-with-terraform) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented on tests/post/steps/test_salt_api.py:
```diff
@@ -48,7 +48,8 @@ def login_salt_api(host, username, password, version, context, request):
         cmd_output = host.check_output(cmd_ip)
-    ip = json.loads(cmd_output)['local'][0]
+    # We take the last IP, otherwise we can have the control plane VIP instead
+    ip = json.loads(cmd_output)['local'][-1]
```
Isn't it the opposite in Vagrant?

alexandre-allard replied:
Just checked, looks like it is...

gdemonet replied:
We should probably replace this whole way of retrieving the control-plane IP and use the grain `metalk8s.control_plane_ip` instead (`metalk8s.node.grains` already handles ignoring any apiserver VIP configured)

alexandre-allard replied:
smart

alexandre-allard committed:
tf: raise resources for bootstrap node

Change the openstack_flavour_name var type into a
map to specify flavour by instance type
(bastion, bootstrap & nodes).
Use a m1.large flavour for bootstrap node
instead of a m1.medium because machine consumes
all its memory and become less responsive.

bert-e commented:
# History mismatch



Merge commit #a8d358d1cac49ed0512a3d8cde85e9c415045688 on the integration branch
`w/2.1/improvement/deploy-metalk8s-with-terraform` is merging a branch which is neither the current
branch `improvement/deploy-metalk8s-with-terraform` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/deploy-metalk8s-with-terraform` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
tests: get control plane ip from grain

Retrieve control plane IP from grains
metalk8s.control_plane_ip grain instead
of getting the first IP configured on the
controle plane network interface.
With the introduction of the VIP in terraform
the old behavior could lead to retrieving the
VIP address instead of the real IP of the interface.

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/improvement/deploy-metalk8s-with-terraform` with contents from `w/2.3/improvement/deploy-metalk8s-with-terraform`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/improvement/deploy-metalk8s-with-terraform origin/development/2.4
 $ git merge origin/w/2.3/improvement/deploy-metalk8s-with-terraform
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/deploy-metalk8s-with-terraform
```

bert-e commented:
# History mismatch



Merge commit #9ebeb8598fc5e1c8a26ea12679c14fe85859c086 on the integration branch
`w/2.1/improvement/deploy-metalk8s-with-terraform` is merging a branch which is neither the current
branch `improvement/deploy-metalk8s-with-terraform` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/deploy-metalk8s-with-terraform` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/improvement/deploy-metalk8s-with-terraform` with contents from `w/2.3/improvement/deploy-metalk8s-with-terraform`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/improvement/deploy-metalk8s-with-terraform origin/development/2.4
 $ git merge origin/w/2.3/improvement/deploy-metalk8s-with-terraform
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/deploy-metalk8s-with-terraform
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/deploy-metalk8s-with-terraform` with contents from `w/2.4/improvement/deploy-metalk8s-with-terraform`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/deploy-metalk8s-with-terraform origin/development/2.5
 $ git merge origin/w/2.4/improvement/deploy-metalk8s-with-terraform
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/deploy-metalk8s-with-terraform
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.4/improvement/deploy-metalk8s-with-terraform` with contents from `w/2.3/improvement/deploy-metalk8s-with-terraform`
and `development/2.4`.

Please resolve the conflict on **the integration branch** (`w/2.4/improvement/deploy-metalk8s-with-terraform`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.4/improvement/deploy-metalk8s-with-terraform
 $ git pull  # or "git reset --hard origin/w/2.4/improvement/deploy-metalk8s-with-terraform"
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/w/2.3/improvement/deploy-metalk8s-with-terraform
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/deploy-metalk8s-with-terraform
```
